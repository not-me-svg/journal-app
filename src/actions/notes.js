import Swal from 'sweetalert2'
import { addDoc, db, collection, deleteDoc, doc, updateDoc } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const docRef = await addDoc(collection(db, `${ uid }`, "journal/notes"), newNote);

            dispatch(activateNote(docRef.id, newNote));
            dispatch(addNewNote(docRef.id, newNote));
        } catch (error) {
            console.log(error);
        }
    }
}

export const activateNote = (id, note) => {
    return {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
}

export const addNewNote = (id, note) => {
    return {
        type: types.notesAddNew,
        payload: {
            id, 
            ...note
        }
    }
}

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        }

        const noteToFirestore = { ...note }
        delete noteToFirestore.id;

        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        await updateDoc(noteRef, noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    }
}

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: {
            id,
            note: {
                id,
                ...note
            }
        }
    }
}

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active:activeNote } = getState().notes;
        
        Swal.fire({
            title: "Uploading...",
            text: "Please wait...",
            showConfirmButton: false,
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
        
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        const noteRef = doc(db,`${uid}/journal/notes/${id}`);
        await deleteDoc(noteRef);

        dispatch( deleteNote(id));
    }
}

export const deleteNote = (id) => {
    return {
        type: types.notesDelete,
        payload: id
    }
}

export const noteLogout = () => {
    return {
        type: types.notesLogOutCleaning
    }
}