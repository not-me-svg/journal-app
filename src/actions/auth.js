import Swal from 'sweetalert2'
import { types } from "../types/types";
import { startLoading, finishLoading } from "./ui";
import { noteLogout } from './notes';
import { 
    auth,
    createUserWithEmailAndPassword,
    googleAuthProvider,
    signInWithEmailAndPassword,
    signInWithPopup,
    signOut,
    updateProfile
} from "../firebase/firebase-config";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch( startLoading() );

        signInWithEmailAndPassword(auth, email, password)
            .then(({ user }) => {
                dispatch( login(user.uid, user.displayName) );

                dispatch( finishLoading() );
            }).catch(err => {
                console.log(err)
                dispatch( finishLoading() );
                Swal.fire('Error', err.message, 'error');
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        signInWithPopup(auth, googleAuthProvider)
            .then(({ user }) => {
                dispatch( login(user.uid, user.displayName) )
            }).catch(err => {
                console.log(err)
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name })

                dispatch( login(user.uid, user.displayName) )
            }).catch(err => {
                console.log(err)
                Swal.fire('Error', err.message, 'error');
            })
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogOut = () => {
    return async (dispatch) => {
        await signOut(auth);

        dispatch(logout());
        dispatch(noteLogout());
    }
}

export const logout = () => ({
    type: types.logout
})