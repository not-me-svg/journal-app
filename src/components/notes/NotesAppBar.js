import React from 'react';
import { useSelector, useDispatch } from "react-redux";
import { startDeleteNote, startSaveNote } from '../../actions/notes';

export const NotesAppBar = () => {

  const dispatch = useDispatch();
  const { active } = useSelector( state => state.notes );

  const handleSave = () => {
    dispatch(startSaveNote(active));
  }

  const handleDelete = () => {
    dispatch(startDeleteNote(active.id));
  }

  return (
    <div className="ja__notes-bar w-100 d-flex jc-space-between ai-center">
        <span>28 de agosto de 2020</span>

        <div>
            <button
              className="btn btn--secondary mr--sm" 
              onClick={ handleSave }>Save</button>

            <button
              className="btn btn--secondary" 
              onClick={ handleDelete }>Delete</button>
        </div>
    </div>
  );
};
