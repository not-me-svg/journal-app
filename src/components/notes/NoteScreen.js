import React, { useRef, useEffect } from 'react';
import { useSelector, useDispatch } from "react-redux";

import { NotesAppBar } from './NotesAppBar';
import { useForm } from '../../hooks/useForm';
import { activateNote, startUploading } from '../../actions/notes';

export const NoteScreen = () => {

  const dispatch = useDispatch();
  const { active:note } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm(note);
  const { body, title, id } = formValues;
  
  const activeId = useRef( note.id );
 
  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);

      activeId.current = note.id;
    }
  }, [note, reset]);
  
  useEffect(() => {
    dispatch(activateNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const handlePictureClick = (e) => {
    e.preventDefault();
    document.querySelector('#file-selector').click()
  }

  const handleFileChange = (e) => {
    const file = e.target.files[0];

    if (file) {
      dispatch(startUploading(file));
    }
  }

  return (
    <div>
      <NotesAppBar { ...note }/>

      <div className="ja__note-screen">
        <form>
          <div class="d-flex flex-column">
            <input 
              className="input"
              type="text"
              placeholder="Note title"
              autoComplete="off"
              name="title"
              value={ title }
              onChange={ handleInputChange } />

            <br />

            <textarea 
              className="textarea"
              placeholder="xxxx" 
              name="Note body" 
              id="" 
              cols="30" 
              rows="10"
              name="body"
              value={ body }
              onChange={ handleInputChange } ></textarea>
              <br />

            {
              (note.url)
              ?
              (
                <figure>
                  <img 
                      src={ note.url }
                      alt="XXX"
                      className="w-100 h-100"/>
                </figure>
              )
              :
              (
                <div>
                  <button 
                    className="btn btn--secondary mr--sm" 
                    onClick={ handlePictureClick }>Upload Picture</button>
                </div>
              )
            }
          </div>
        </form>
      </div>

      <input 
        type="file" 
        onChange={ handleFileChange }
        id="file-selector"
        name="file"
        className="hide" />
    </div>
  );
};
