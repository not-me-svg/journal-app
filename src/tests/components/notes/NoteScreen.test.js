import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { activateNote } from '../../../actions/notes';
import { NoteScreen } from '../../../components/notes/NoteScreen';


jest.mock('../../../actions/notes', () => ({
    activateNote: jest.fn(),
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {
        uid: '1',
        name: 'Sample'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: 1234,
            title: 'Hello',
            body: 'friend',
            date: 0
        },
        notes: []
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <NoteScreen /> 
    </Provider>

)

describe('Test in <NoteScreen />', () => {
    test('Should show', () => {
        expect( wrapper ).toMatchSnapshot();
    })

    test('Should trigger active note', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Hello friend'
            }
        });

        expect( activateNote ).toHaveBeenLastCalledWith(
            1234,
            {
                body: 'friend',
                title: 'Hello friend',
                id: 1234,
                date: 0
            }
        );
    }) 
})
