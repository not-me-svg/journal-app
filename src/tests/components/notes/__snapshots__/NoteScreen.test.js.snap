// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> debe de mostrarse correctamente 1`] = `ReactWrapper {}`;

exports[`Test in <NoteScreen /> Should show 1`] = `ReactWrapper {}`;
